solver/src/driver.rs

```diff
@@ -32,16 +34,25 @@ impl Driver {
                 None => return Ok(()),
                 Some(settlement) => settlement,
             };
-        // TODO: encode settlement, check if we need to approve spending to uniswap
+        // TODO: check if we need to approve spending to uniswap
         // TODO: use retry transaction sending crate for updating gas prices
-        self.contract
-            .settle_call(&settlement)
-            .await
-            .context("settle call failed")?;
-        self.contract
-            .settle_send(&settlement)
-            .await
-            .context("settle send failed")?;
+        let settle = || {
+            self.settlement.settle(
+                settlement.tokens(),
+                settlement.clearing_prices(),
+                settlement.fee_factor,
+                settlement
+                    .encode_trades()
+                    .expect("naive solver created invalid settlement"),
+                settlement.encode_interactions(
+                    &self.uniswap_router.address(),
+                    &self.settlement.address(),
```

Comment:
Instead of passing in all dependencies for all possible interactions into this function, I'd like to see each interaction to somehow know intrinsically (via a Factory or the like) how it can encode itself and what dependencies it needs. 
While this is fine for the demo with just one interaction, I think making sure we can easily add new interactions into the codebase is one of the first refactorings we should look at before accruing more debt.